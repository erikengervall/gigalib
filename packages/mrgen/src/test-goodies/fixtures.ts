import { MRGenOptions } from '../types/mrgen-options';

export const createMockMRGenOptions = (
  overrides?: Partial<MRGenOptions>,
): { mrGenOptions: MRGenOptions } => {
  return {
    mrGenOptions: {
      data: {
        queryParams: true,
        payload: true,
        params: true,
        responseBody: true,
      },

      httpMethod: 'POST',

      shared: {
        apiDir: 'packages/shared/src/api',
        file: 'packages/shared/src/api/v1/user/get.ts',
        buildCommand: 'yarn workspace shared ci:build',
      },
      backend: {
        file: 'packages/backend/src/handlers/v1/user/get-handler.ts',
      },
      frontend: {
        file: 'packages/frontend1/src/api/hooks/v1-user-get.ts',
      },
      ...overrides,
    },
  };
};
