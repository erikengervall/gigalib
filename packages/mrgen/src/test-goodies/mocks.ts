import cp from 'node:child_process';
import fs from 'node:fs';

export const mockExecSync = cp.execSync as jest.Mock;
export const mockMkdirSync = fs.mkdirSync as jest.Mock;
export const mockWriteFileSync = fs.writeFileSync as jest.Mock;
