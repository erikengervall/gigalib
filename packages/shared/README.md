# @engervall/shared

Utilities and stuff.
