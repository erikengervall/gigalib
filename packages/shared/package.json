{
  "name": "@engervall/shared",
  "version": "1.0.6-alpha.0",
  "main": "dist/index.js",
  "repository": "git@github.com:erikengervall/gigalib.git",
  "author": "Erik Engervall <erik.engervall@gmail.com>",
  "scripts": {
    "----- CI": "",
    "ci:build": "tsc",
    "ci:format": "prettier --check './src/**/*.{ts,tsx}' --ignore-path ../../.prettierignore",
    "ci:lint": "eslint 'src/**/*.{ts,tsx}'",
    "ci:test:unit": "jest",
    "----- LOCAL DEV": "",
    "dev": "rm -rf dist && tsc --watch",
    "----- MISC": "",
    "format:write": "prettier --write './src/**/*.{ts,tsx}'",
    "clean": "rm -rf dist",
    "reinstall": "yarn clean && rm -rf node_modules"
  },
  "dependencies": {
    "qs": "^6.11.2",
    "zod": "^3.21.4",
    "zod-validation-error": "^1.3.1"
  }
}
