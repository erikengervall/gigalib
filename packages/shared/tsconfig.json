{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "declaration": true,
    "lib": ["es2021"],
    "module": "CommonJS",
    "noEmit": false,
    "sourceMap": true,
    "target": "ES6",

    "baseUrl": "src",
    "outDir": "dist",
    "rootDir": "src"
  },
  "include": ["src", "../../scripts/change-case.ts"],
  "exclude": [
    "_warmup",
    ".cache",
    ".next",
    ".serverless",
    ".vscode",
    ".vscode",
    ".webpack",
    "*.dist*",
    "*.spec*",
    "*.test*",
    "build",
    "coverage",
    "cypress-coverage",
    "dist",
    "jest-coverage",
    "node_modules"
  ]
}
