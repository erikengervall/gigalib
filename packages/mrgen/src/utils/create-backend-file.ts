import { SHARED_EXPORTS } from './create-shared-file';

import { MRGenOptions } from '../types/mrgen-options';
import { camelCase } from './change-case';
import { createFileWithContent } from './create-file-with-content';
import { fileName } from './file-name';
import { formatFile } from './format-file';
import { getAPIType } from './get-api-type';

export function createBackendFile(options: MRGenOptions) {
  const { APIType } = getAPIType(options);

  /**
   * '/Users/.../<monorepo>/packages/backend/src/handlers/v1/user/get-handler.ts'
   * -> get-handler
   */
  const { withoutExt } = fileName(options.backend.file);
  const handlerName = camelCase(withoutExt);

  const payloadContent = options.data.payload
    ? `
const parsed${SHARED_EXPORTS.Payload} = ${APIType.payload}.safeParse(req.body);
if (!parsed${SHARED_EXPORTS.Payload}.success) {
  res.status(400).json(); // TODO: Include error details.
  return;
}`
    : '';

  const queryParamsContent = options.data.queryParams
    ? `
  const parsed${SHARED_EXPORTS.QueryParams} = ${APIType.queryParams}.safeParse(req.query);
  if (!parsed${SHARED_EXPORTS.QueryParams}.success) {
    res.status(400).json(); // TODO: Include error details.
    return;
  }`
    : '';

  const paramsContent = options.data.params
    ? `
  const parsed${SHARED_EXPORTS.Params} = ${APIType.params}.safeParse(req.params);
  if (!parsed${SHARED_EXPORTS.Params}.success) {
    res.status(400).json(); // TODO: Include error details.
    return;
  }`
    : '';

  const responseBodyContent = options.data.responseBody
    ? `
const ${camelCase(SHARED_EXPORTS.ResponseBody)}: ${APIType.responseBody} = {
  data: {},
};
return res.status(200).json(responseBody);
    `
    : 'return res.status(200).send();';

  const backendFileContent = `
  import { API } from 'shared';
  import type { RequestHandler } from 'express';
  
  export const ${handlerName}: RequestHandler = async (req, res) => {
    ${payloadContent}

    ${queryParamsContent}

    ${paramsContent}
  
    ${responseBodyContent}
  }
  `;

  createFileWithContent(options.backend.file, backendFileContent);

  formatFile(options.backend.file);
}
