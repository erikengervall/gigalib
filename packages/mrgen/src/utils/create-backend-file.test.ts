import cp from 'node:child_process';
import fs from 'node:fs';

import { logSnapshots } from '@engervall/shared';

import { createMockMRGenOptions } from '../test-goodies/fixtures';
import { createBackendFile } from './create-backend-file';

it('should create and format backend files', async () => {
  // Arrange
  const { mrGenOptions } = createMockMRGenOptions();

  // Act
  createBackendFile(mrGenOptions);

  // Assert
  expect((fs.mkdirSync as jest.Mock).mock.calls).toMatchSnapshot(
    'fs.mkdirSync',
  );
  expect((fs.writeFileSync as jest.Mock).mock.calls).toMatchSnapshot(
    'fs.writeFileSync',
  );
  expect((cp.execSync as jest.Mock).mock.calls).toMatchSnapshot('cp.execSync');
  logSnapshots();
});
