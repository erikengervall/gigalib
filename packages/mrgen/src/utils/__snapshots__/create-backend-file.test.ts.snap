// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create and format backend files: aggregateLogProxySpy 1`] = `
[
  {
    "debug": [
      "mkdirSync",
      "packages/backend/src/handlers/v1/user",
    ],
  },
  {
    "debug": [
      "writeFileSync",
      "packages/backend/src/handlers/v1/user/get-handler.ts",
    ],
  },
  {
    "debug": [
      "Running command",
      "yarn prettier --write packages/backend/src/handlers/v1/user/get-handler.ts",
    ],
  },
]
`;

exports[`should create and format backend files: cp.execSync 1`] = `
[
  [
    "yarn prettier --write packages/backend/src/handlers/v1/user/get-handler.ts",
  ],
]
`;

exports[`should create and format backend files: fs.mkdirSync 1`] = `
[
  [
    "packages/backend/src/handlers/v1/user",
    {
      "recursive": true,
    },
  ],
]
`;

exports[`should create and format backend files: fs.writeFileSync 1`] = `
[
  [
    "packages/backend/src/handlers/v1/user/get-handler.ts",
    "
  import { API } from 'shared';
  import type { RequestHandler } from 'express';
  
  export const getHandler: RequestHandler = async (req, res) => {
    
const parsedPayload = API.V1.User.Get.Payload.safeParse(req.body);
if (!parsedPayload.success) {
  res.status(400).json(); // TODO: Include error details.
  return;
}

    
  const parsedQueryParams = API.V1.User.Get.QueryParams.safeParse(req.query);
  if (!parsedQueryParams.success) {
    res.status(400).json(); // TODO: Include error details.
    return;
  }

    
  const parsedParams = API.V1.User.Get.Params.safeParse(req.params);
  if (!parsedParams.success) {
    res.status(400).json(); // TODO: Include error details.
    return;
  }
  
    
const responseBody: API.V1.User.Get.ResponseBody = {
  data: {},
};
return res.status(200).json(responseBody);
    
  }
  ",
    "utf-8",
  ],
]
`;
